### PR TITLE
CBG-3516: Release sequences in ranges if allocating a sequence lower than a previously seen sequence

### PR DESCRIPTION
**Pull request by bbrks**
CBG-3516

Writes a test to reproduce the scenario where the `_sync:seq` doc has been lowered, or we're seeing documents being updated that came from different clusters (with higher sequence numbers)

Existing behaviour was to release sequences all the way from the current sequence up to the new high sequence - one at a time. This PR changes that to release sequences in ranges, wherever possible, but retains the existing functional behaviour.

Added a warning if the sequence gap is much larger than expected (`100 * maxBatchSize`) - there are situations where other SG nodes have reserved a batch of sequences and haven't used them yet.

## Pre-review checklist
- [ ] Removed debug logging (`fmt.Print`, `log.Print`, ...)
- [ ] Logging sensitive data? Make sure it's tagged (e.g. `base.UD(docID)`, `base.MD(dbName)`)
- [ ] Updated relevant information in the API specifications (such as endpoint descriptions, schemas, ...) in `docs/api`

## [Integration Tests](https://jenkins.sgwdev.com/job/SyncGateway-Integration/build?delay=0sec)
- [ ] `GSI=true,xattrs=true` https://jenkins.sgwdev.com/job/SyncGateway-Integration/2128/

### CODE REVIEW

File: db/sequence_allocator.go [hunk 1]
```diff
@@ -273,3 +344,56 @@ func (s *sequenceAllocator) waitForReleasedSequences(ctx context.Context, startT
...
+				break
+			}
+		}
+	}
+
+	s.mutex.Unlock()
+
+	err = s.releaseSequences(ctx, sequencesToRelease)
+	if err != nil {
+		s.mutex.Unlock()
```
Comment (torcolvin) [inline db/sequence_allocator.go]: this unlock doesn't seem right

Comment (bbrks) [reply db/sequence_allocator.go]: You're right - thanks!

File: db/sequence_allocator.go [hunk 2]
```diff
@@ -273,3 +344,55 @@ func (s *sequenceAllocator) waitForReleasedSequences(ctx context.Context, startT
...
+	sequenceGapTolerance := uint64(maxBatchSize * 100)
+	if existingSequence >= sequence+sequenceGapTolerance {
+		base.ErrorfCtx(ctx, "Existing sequence (%d) was much higher than the one allocated (%d) - document sequencing considered corrupt. Should run resync with regenerate_sequences.", existingSequence, s.last)
+	}
+
+	// we should fill the gaps with unused sequences, just in case SG nodes aren't actually going to release these (if the gap was for a valid reason)
+	if existingSequence >= sequence {
+		for {
+			// keep a record of which sequences we should release as unused later
+			sequencesToRelease = append(sequencesToRelease, sequence)
```
Comment (bbrks) [inline db/sequence_allocator.go:368]: Don't store all sequences in slice - grow ranges inside this slice

File: db/sequence_allocator.go [hunk 3]
```diff
@@ -185,6 +179,18 @@ func (s *sequenceAllocator) nextSequence(ctx context.Context) (sequence uint64,
 	return sequence, nil
 }
 
+func (s *sequenceAllocator) _nextSequence(ctx context.Context) (sequence uint64, sequencesReserved bool, err error) {
+	if s.last >= s.max {
+		if err := s._reserveSequenceRange(ctx); err != nil {
```
Comment (bbrks) [inline db/sequence_allocator.go:184]: Have a way to reserve a sequence range from current sequence up to the high doc.Sequence

File: db/crud_test.go [hunk 4]
```diff
@@ -1628,3 +1629,51 @@ func TestPutStampClusterUUID(t *testing.T) {
...
+	t.Logf("doc sequence: %d", doc.Sequence)
+
+	// wait for the doc to be received
+	err = db.changeCache.waitForSequence(ctx, doc.Sequence, time.Second*30)
+	require.NoError(t, err)
+
+	expectedReleasedSequenceCount := numSequences + 1 // 1 sequence allocated for the first doc write
+	releasedSequenceCount := db.DbStats.Database().SequenceReleasedCount.Value() - startReleasedSequenceCount
+	assert.Equal(t, int64(expectedReleasedSequenceCount), releasedSequenceCount)
+}
```
Comment (bbrks) [inline db/crud_test.go:1679]: disable sequence batching and check for range doc to assert that we're not writing one doc per released sequence?

**Comment by adamcfraser** [comment]
Merged with https://github.com/couchbase/sync_gateway/commit/20e5e476382fddae034ec7a3961e0d14a862ce9b

**Pull request closed without merging**